<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<html>
<!-- This is the manual of MafFilter, version 1.3.1.

Copyright (C) 2018 Julien Y. Dutheil -->
<!-- Created by GNU Texinfo 6.1, http://www.gnu.org/software/texinfo/ -->
<head>
<title>MafFilter Manual 1.3.1: DistanceEstimation</title>

<meta name="description" content="MafFilter Manual 1.3.1: DistanceEstimation">
<meta name="keywords" content="MafFilter Manual 1.3.1: DistanceEstimation">
<meta name="resource-type" content="document">
<meta name="distribution" content="global">
<meta name="Generator" content="makeinfo">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
<link href="index.html#Top" rel="start" title="Top">
<link href="index.html#SEC_Contents" rel="contents" title="Table of Contents">
<link href="Phylogenetics.html#Phylogenetics" rel="up" title="Phylogenetics">
<link href="DistanceBasedPhylogeny.html#DistanceBasedPhylogeny" rel="next" title="DistanceBasedPhylogeny">
<link href="Phylogenetics.html#Phylogenetics" rel="prev" title="Phylogenetics">
<style type="text/css">
<!--
a.summary-letter {text-decoration: none}
blockquote.indentedblock {margin-right: 0em}
blockquote.smallindentedblock {margin-right: 0em; font-size: smaller}
blockquote.smallquotation {font-size: smaller}
div.display {margin-left: 3.2em}
div.example {margin-left: 3.2em}
div.lisp {margin-left: 3.2em}
div.smalldisplay {margin-left: 3.2em}
div.smallexample {margin-left: 3.2em}
div.smalllisp {margin-left: 3.2em}
kbd {font-style: oblique}
pre.display {font-family: inherit}
pre.format {font-family: inherit}
pre.menu-comment {font-family: serif}
pre.menu-preformatted {font-family: serif}
pre.smalldisplay {font-family: inherit; font-size: smaller}
pre.smallexample {font-size: smaller}
pre.smallformat {font-family: inherit; font-size: smaller}
pre.smalllisp {font-size: smaller}
span.nolinebreak {white-space: nowrap}
span.roman {font-family: initial; font-weight: normal}
span.sansserif {font-family: sans-serif; font-weight: normal}
ul.no-bullet {list-style: none}
-->
</style>
<link rel="stylesheet" type="text/css" href="http://www.w3.org/StyleSheets/Core/Steely">


</head>

<body lang="en">
<a name="DistanceEstimation"></a>
<div class="header">
<p>
Next: <a href="DistanceBasedPhylogeny.html#DistanceBasedPhylogeny" accesskey="n" rel="next">DistanceBasedPhylogeny</a>, Previous: <a href="Phylogenetics.html#Phylogenetics" accesskey="p" rel="prev">Phylogenetics</a>, Up: <a href="Phylogenetics.html#Phylogenetics" accesskey="u" rel="up">Phylogenetics</a> &nbsp; [<a href="index.html#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<a name="Distance-matrix-estimation"></a>
<h4 class="subsubsection">2.3.2.1 Distance matrix estimation</h4>

<p>Estimates a pairwise distance matrix.
</p>
<a name="Synopsis_003a-29"></a>
<h3 class="heading">Synopsis:</h3>

<table class="cartouche" border="1"><tr><td>
<div class="example">
<pre class="example">maf.filter=                                 \
    [...],
    DistanceEstimation(                     \
        method=count,                       \
        gap_option=no_double_gap,           \
        unresolved_as_gap=no,               \
        extended_names=yes),                \
    [...]
</pre></div>
<p>or
</p><div class="example">
<pre class="example">maf.filter=                                 \
    [...],
    DistanceEstimation(                     \
        method=ml,                          \
        model=K80(kappa=2),                 \
        rate=Gamma(n=4, alpha=0.5),         \
        parameter_estimation=initial,       \
        max_freq_gaps=0.33,                 \
        gaps_as_unresolved=yes,             \
        profiler=none,                      \
        message_handler=none,               \
        extended_names=yes),                \
    [...]
</pre></div>
</td></tr></table>

<a name="Arguments_003a-29"></a>
<h3 class="heading">Arguments:</h3>

<dl compact="compact">
<dt><code>method={count|ml}</code></dt>
<dd><p>Method used to estimate distance, either observed count or maximum likelihood estimate.
</p></dd>
</dl>

<p>Further arguments for the observed counts:
</p><dl compact="compact">
<dt><code>gap_option={string}</code></dt>
<dd><p>Specifies how to deal with gaps:
</p><dl compact="compact">
<dt><code>all</code></dt>
<dd><p>All positions are used. Gaps are considered as a fifth character.
</p></dd>
<dt><code>no_full_gap</code></dt>
<dd><p>Positions only made of gaps in the alignment block are ignored. Alternatively, a gap in the two sequences is ocnsidered as a match (gap are a &ldquo;fifth&rdquo; charcater).
</p></dd>
<dt><code>no_double_gap</code></dt>
<dd><p>For each pairwise comparison, positions where a gap is found in both sequences are ignored.
</p></dd>
<dt><code>no_gap</code></dt>
<dd><p>For each pairwise comparison, any gap-containing position is ignored. This is the recommended option for building phylogenies.
</p></dd>
</dl>

</dd>
<dt><code>unresolved_as_gap={yes|no}</code></dt>
<dd><p>Tell is unresolved characters should be treated as gaps (usually in order to be ignored).
</p>
</dd>
<dt><code>extended_names={boolean}</code></dt>
<dd><p>Tell if sequence coordinates should be included in the sequence names stored in the output matrix.
</p></dd>
</dl>

<p>Further arguments for the ML method:
</p><dl compact="compact">
<dt><code>model={substitution model description}</code></dt>
<dd><p>See the Bio++ Program Suite manual for a description of substitution models available. Only nucleotide models can be used.
</p><div class="example">
<pre class="example">model=JC
model=K80(kappa=2)
model=T92(kappa=2, theta=0.5)
model=GTR
</pre></div>

</dd>
<dt><code>rate={rate distribution description}</code></dt>
<dd><p>See the Bio++ Program Suite manual for a description of available options.
</p><div class="example">
<pre class="example">rate=Constant
rate=Gamma(n=4, alpha=0.5)
</pre></div>

</dd>
<dt><code>profiler={none|std|{path}}</code></dt>
<dd><p>Where to print optimization steps (nowhere, standard output or to a given file).
</p>
</dd>
<dt><code>message_handler={none|std|{path}}</code></dt>
<dd><p>Where to log optimization (nowhere, standard output or to a given file).
</p>
</dd>
<dt><code>parameter_estimation={initial|pairwise}</code></dt>
<dd><p>How to estimate substitution process parameters (for instance kappa and alpha).
Available options are either to leave them to their initial values, or to estimate them for each pair of sequences.
</p>
</dd>
<dt><code>max_freq_gaps={float}</code></dt>
<dd><p>The maximum proportion of gaps for a site to be included in the analysis.
</p>
</dd>
<dt><code>gaps_as_unresolved={yes/no}</code></dt>
<dd><p>Tell if remaining gaps should be converted to &rsquo;N&rsquo; before likelihood computation. This should be &rsquo;yes&rsquo; unless you specify a substitution model which explicitely allows for gaps.
</p></dd>
</dl>


<a name="Extra_002ddata_003a"></a>
<h3 class="heading">Extra-data:</h3>

<dl compact="compact">
<dt><code>CountDistance or MLDistance</code></dt>
<dd><p>The estimated pairwise distance matrix.
</p></dd>
</dl>





</body>
</html>
