<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<html>
<!-- This is the manual of MafFilter, version 1.3.1.

Copyright (C) 2018 Julien Y. Dutheil -->
<!-- Created by GNU Texinfo 6.1, http://www.gnu.org/software/texinfo/ -->
<head>
<title>MafFilter Manual 1.3.1: General</title>

<meta name="description" content="MafFilter Manual 1.3.1: General">
<meta name="keywords" content="MafFilter Manual 1.3.1: General">
<meta name="resource-type" content="document">
<meta name="distribution" content="global">
<meta name="Generator" content="makeinfo">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
<link href="index.html#Top" rel="start" title="Top">
<link href="index.html#SEC_Contents" rel="contents" title="Table of Contents">
<link href="Introduction.html#Introduction" rel="up" title="Introduction">
<link href="Filters.html#Filters" rel="next" title="Filters">
<link href="Run.html#Run" rel="prev" title="Run">
<style type="text/css">
<!--
a.summary-letter {text-decoration: none}
blockquote.indentedblock {margin-right: 0em}
blockquote.smallindentedblock {margin-right: 0em; font-size: smaller}
blockquote.smallquotation {font-size: smaller}
div.display {margin-left: 3.2em}
div.example {margin-left: 3.2em}
div.lisp {margin-left: 3.2em}
div.smalldisplay {margin-left: 3.2em}
div.smallexample {margin-left: 3.2em}
div.smalllisp {margin-left: 3.2em}
kbd {font-style: oblique}
pre.display {font-family: inherit}
pre.format {font-family: inherit}
pre.menu-comment {font-family: serif}
pre.menu-preformatted {font-family: serif}
pre.smalldisplay {font-family: inherit; font-size: smaller}
pre.smallexample {font-size: smaller}
pre.smallformat {font-family: inherit; font-size: smaller}
pre.smalllisp {font-size: smaller}
span.nolinebreak {white-space: nowrap}
span.roman {font-family: initial; font-weight: normal}
span.sansserif {font-family: sans-serif; font-weight: normal}
ul.no-bullet {list-style: none}
-->
</style>
<link rel="stylesheet" type="text/css" href="http://www.w3.org/StyleSheets/Core/Steely">


</head>

<body lang="en">
<a name="General"></a>
<div class="header">
<p>
Previous: <a href="Run.html#Run" accesskey="p" rel="prev">Run</a>, Up: <a href="Introduction.html#Introduction" accesskey="u" rel="up">Introduction</a> &nbsp; [<a href="index.html#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<a name="General-options"></a>
<h3 class="section">1.3 General options</h3>

<a name="Synopsis_003a"></a>
<h3 class="heading">Synopsis:</h3>

<table class="cartouche" border="1"><tr><td>
<div class="example">
<pre class="example">input.file=mydata.maf.gz
input.file.compression=gzip
input.format=Maf
output.log=mydata.maffilter.log
...
</pre></div>

<p>or
</p><div class="example">
<pre class="example">input.file=mydata.fasta
input.file.compression=none
input.format=Fasta(zero_based=yes)
output.log=mydata.maffilter.log
...
</pre></div>
</td></tr></table>

<a name="Detailed-description-of-commands_003a"></a>
<h3 class="heading">Detailed description of commands:</h3>

<dl compact="compact">
<dt><code>input.file={path}</code></dt>
<dd><p>The MAF file to parse
</p>
</dd>
<dt><code>input.file.compression=[none|zip|gzip|bzip2]</code></dt>
<dd><p>The compression format of the input file.
</p>
</dd>
<dt><code>input.format={Maf|Fasta}</code></dt>
<dd><p>MafFilter takes by default a MAF file. It is, however, possible to use a fasta file as input,
which will lead MafFilter to run in &quot;single sequence&quot; mode.
In this mode, each sequence of the original file is considered as a distinct block.
This mode might be useful to extract annotations and estimate simple statistics.
Fasta files can be compressed in the same way as MAF files.
The sequence names are expected to be of the form Species:Chromosome:start:strand:length, with possible empty fields.
Coordinates are expected to be 0-based (that is, the first nucleotide is position 0). However, passing the <code>zero_based=no</code> argument to the format name will assume that the first nucleotide is position 1.
</p>
</dd>
<dt><code>input.dots={error|as_gaps|as_unresolved}</code></dt>
<dd><p>If set to <code>as_gaps</code>, dotted characters in sequences will be converted to gaps and treated as such in any further step of the analysis, including output to files. If set to <code>as_unresolved</code>, dots will be converted to &rsquo;N&rsquo;.
</p>
</dd>
<dt><code>input.check_sequence_size={boolean}</code></dt>
<dd><p>Tell if sequence sizes are consistent (default: true). An error will be returned in case of inconsistency.
Desabling size check can be useful in combination with input.dots=as_gaps. In case of inconsistency, a warning will be displayed unless verbose is set to false.
</p>
</dd>
<dt><code>output.log={path}</code></dt>
<dd><p>The file where to write log messages.
</p>
</dd>
<dt><code>maf.filter={Filter1, Filter2, ...}</code></dt>
<dd><p>The main command, specifying the list of filter commands to apply, in the given order (that is, the input of filter 2 will be the output of filter 1).
Filters take a series of arguments, specified within parentheses: <code>Filter(arg1=opt1, arg2=opt2, etc)</code>.
The next section present all available filters and their corresponding arguments.
</p></dd>
</dl>





</body>
</html>
