<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<html>
<!-- This is the manual of MafFilter, version 1.3.1.

Copyright (C) 2018 Julien Y. Dutheil -->
<!-- Created by GNU Texinfo 6.1, http://www.gnu.org/software/texinfo/ -->
<head>
<title>MafFilter Manual 1.3.1: Top</title>

<meta name="description" content="MafFilter Manual 1.3.1: Top">
<meta name="keywords" content="MafFilter Manual 1.3.1: Top">
<meta name="resource-type" content="document">
<meta name="distribution" content="global">
<meta name="Generator" content="makeinfo">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
<link href="#Top" rel="start" title="Top">
<link href="#SEC_Contents" rel="contents" title="Table of Contents">
<link href="../dir/index.html" rel="up" title="(dir)">
<link href="Introduction.html#Introduction" rel="next" title="Introduction">
<style type="text/css">
<!--
a.summary-letter {text-decoration: none}
blockquote.indentedblock {margin-right: 0em}
blockquote.smallindentedblock {margin-right: 0em; font-size: smaller}
blockquote.smallquotation {font-size: smaller}
div.display {margin-left: 3.2em}
div.example {margin-left: 3.2em}
div.lisp {margin-left: 3.2em}
div.smalldisplay {margin-left: 3.2em}
div.smallexample {margin-left: 3.2em}
div.smalllisp {margin-left: 3.2em}
kbd {font-style: oblique}
pre.display {font-family: inherit}
pre.format {font-family: inherit}
pre.menu-comment {font-family: serif}
pre.menu-preformatted {font-family: serif}
pre.smalldisplay {font-family: inherit; font-size: smaller}
pre.smallexample {font-size: smaller}
pre.smallformat {font-family: inherit; font-size: smaller}
pre.smalllisp {font-size: smaller}
span.nolinebreak {white-space: nowrap}
span.roman {font-family: initial; font-weight: normal}
span.sansserif {font-family: sans-serif; font-weight: normal}
ul.no-bullet {list-style: none}
-->
</style>
<link rel="stylesheet" type="text/css" href="http://www.w3.org/StyleSheets/Core/Steely">


</head>

<body lang="en">
<h1 class="settitle" align="center">MafFilter Manual 1.3.1</h1>




<a name="SEC_Contents"></a>
<h2 class="contents-heading">Table of Contents</h2>

<div class="contents">

<ul class="no-bullet">
  <li><a name="toc-Introduction-1" href="Introduction.html#Introduction">1 Introduction</a>
  <ul class="no-bullet">
    <li><a name="toc-Description-of-the-program" href="Description.html#Description">1.1 Description of the program</a></li>
    <li><a name="toc-How-to-run-the-program" href="Run.html#Run">1.2 How to run the program</a></li>
    <li><a name="toc-General-options" href="General.html#General">1.3 General options</a></li>
  </ul></li>
  <li><a name="toc-Filters-1" href="Filters.html#Filters">2 Filters</a>
  <ul class="no-bullet">
    <li><a name="toc-Extracting-data" href="Extracting.html#Extracting">2.1 Extracting data</a>
    <ul class="no-bullet">
      <li><a name="toc-Subset_003a-remove-sequences-from-blocks-by-species-name" href="Subset.html#Subset">2.1.1 Subset: remove sequences from blocks by species name</a></li>
      <li><a name="toc-SelectOrphans_003a-keep-only-block-with-the-specified-species" href="SelectOrphans.html#SelectOrphans">2.1.2 SelectOrphans: keep only block with the specified species</a></li>
      <li><a name="toc-Merging_003a-merge-synthenic-blocks" href="Merge.html#Merge">2.1.3 Merging: merge synthenic blocks</a></li>
      <li><a name="toc-Concatenating_003a-merge-consecutive-blocks-up-to-a-certain-size" href="Concatenate.html#Concatenate">2.1.4 Concatenating: merge consecutive blocks up to a certain size</a></li>
      <li><a name="toc-OrderFilter_003a-check-and-eventually-discard-unordered-or-overlapping-blocks" href="OrderFilter.html#OrderFilter">2.1.5 OrderFilter: check and eventually discard unordered or overlapping blocks</a></li>
      <li><a name="toc-Extract-features-from-the-alignment" href="ExtractFeature.html#ExtractFeature">2.1.6 Extract features from the alignment</a></li>
      <li><a name="toc-Extract-blocks-for-a-given-chromosome_002fcontig_002fscaffold" href="SelectChr.html#SelectChr">2.1.7 Extract blocks for a given chromosome/contig/scaffold</a></li>
      <li><a name="toc-Split-alignment-blocks-into-smaller-parts" href="WindowSplit.html#WindowSplit">2.1.8 Split alignment blocks into smaller parts</a></li>
    </ul></li>
    <li><a name="toc-Cleaning-alignment-blocks" href="Cleaning.html#Cleaning">2.2 Cleaning alignment blocks</a>
    <ul class="no-bullet">
      <li><a name="toc-Minumum-block-length" href="MinBlockLength.html#MinBlockLength">2.2.1 Minumum block length</a></li>
      <li><a name="toc-Minumum-block-size" href="MinBlockSize.html#MinBlockSize">2.2.2 Minumum block size</a></li>
      <li><a name="toc-Exclude-full-gap-columns" href="XFullGap.html#XFullGap">2.2.3 Exclude full gap columns</a></li>
      <li><a name="toc-Remove-empty-sequences" href="RemoveEmptySequences.html#RemoveEmptySequences">2.2.4 Remove empty sequences</a></li>
      <li><a name="toc-Alignment-filtering" href="AlnFilter.html#AlnFilter">2.2.5 Alignment filtering</a></li>
      <li><a name="toc-Alignment-filtering-2" href="AlnFilter2.html#AlnFilter2">2.2.6 Alignment filtering 2</a></li>
      <li><a name="toc-Entropy-filtering" href="EntropyFilter.html#EntropyFilter">2.2.7 Entropy filtering</a></li>
      <li><a name="toc-Masked-sequences-filtering" href="MaskFilter.html#MaskFilter">2.2.8 Masked sequences filtering</a></li>
      <li><a name="toc-Quality-filtering" href="QualFilter.html#QualFilter">2.2.9 Quality filtering</a></li>
      <li><a name="toc-Remove-sequences-from-features-from-the-alignment" href="FeatureFilter.html#FeatureFilter">2.2.10 Remove sequences from features from the alignment</a></li>
    </ul></li>
    <li><a name="toc-Statistical-analysis" href="Analyzing.html#Analyzing">2.3 Statistical analysis</a>
    <ul class="no-bullet">
      <li><a name="toc-Descriptive-statistics" href="Descriptive.html#Descriptive">2.3.1 Descriptive statistics</a>
      <ul class="no-bullet">
        <li><a name="toc-Size-of-alignment-blocks" href="BlockSize.html#BlockSize">2.3.1.1 Size of alignment blocks</a></li>
        <li><a name="toc-Length-of-alignment-blocks" href="BlockLength.html#BlockLength">2.3.1.2 Length of alignment blocks</a></li>
        <li><a name="toc-Length-of-sequence-for-a-species" href="SequenceLength.html#SequenceLength">2.3.1.3 Length of sequence for a species</a></li>
        <li><a name="toc-Alignment-score" href="AlnScore.html#AlnScore">2.3.1.4 Alignment score</a></li>
        <li><a name="toc-Character-frequencies" href="BlockCounts.html#BlockCounts">2.3.1.5 Character frequencies</a></li>
        <li><a name="toc-Pairwise-divergence" href="PairwiseDivergence.html#PairwiseDivergence">2.3.1.6 Pairwise divergence</a></li>
        <li><a name="toc-Site-statistics" href="SiteStatistics.html#SiteStatistics">2.3.1.7 Site statistics</a></li>
        <li><a name="toc-Counts-of-parsimony-informative-sites-for-a-species-quartet" href="FourSpeciesSitePatternCounts.html#FourSpeciesSitePatternCounts">2.3.1.8 Counts of parsimony informative sites for a species quartet</a></li>
        <li><a name="toc-Site-frequency-spectrum" href="SiteFrequencySpectrum.html#SiteFrequencySpectrum">2.3.1.9 Site frequency spectrum</a></li>
        <li><a name="toc-Polymorphism-statistics" href="PolymorphismStatistics.html#PolymorphismStatistics">2.3.1.10 Polymorphism statistics</a></li>
        <li><a name="toc-Sequence-diversity-statistics" href="DiversityStatistics.html#DiversityStatistics">2.3.1.11 Sequence diversity statistics</a></li>
        <li><a name="toc-Number-of-sequence-clusters_002e" href="CountClusters.html#CountClusters">2.3.1.12 Number of sequence clusters.</a></li>
        <li><a name="toc-Fit-a-substitution-model-and-estimate-parameters" href="ModelFit.html#ModelFit">2.3.1.13 Fit a substitution model and estimate parameters</a></li>
      </ul></li>
      <li><a name="toc-Phylogenetics-1" href="Phylogenetics.html#Phylogenetics">2.3.2 Phylogenetics</a>
      <ul class="no-bullet">
        <li><a name="toc-Distance-matrix-estimation" href="DistanceEstimation.html#DistanceEstimation">2.3.2.1 Distance matrix estimation</a></li>
        <li><a name="toc-Hierarchical-clustering-based-on-distances" href="DistanceBasedPhylogeny.html#DistanceBasedPhylogeny">2.3.2.2 Hierarchical clustering based on distances</a></li>
        <li><a name="toc-Root-trees" href="NewOutgroup.html#NewOutgroup">2.3.2.3 Root trees</a></li>
        <li><a name="toc-Remove-a-species-from-a-tree" href="DropSpecies.html#DropSpecies">2.3.2.4 Remove a species from a tree</a></li>
        <li><a name="toc-Filter-blocks-based-on-their-associated-trees" href="TreeFilter.html#TreeFilter">2.3.2.5 Filter blocks based on their associated trees</a></li>
      </ul></li>
    </ul></li>
    <li><a name="toc-Exporting-blocks-and-data" href="Exporting.html#Exporting">2.4 Exporting blocks and data</a>
    <ul class="no-bullet">
      <li><a name="toc-Write-alignment-blocks-to-a-MAF-file" href="Output.html#Output">2.4.1 Write alignment blocks to a MAF file</a></li>
      <li><a name="toc-Write-alignment-blocks-to-an-external-alignment-file_002e" href="OutputAlignments.html#OutputAlignments">2.4.2 Write alignment blocks to an external alignment file.</a></li>
      <li><a name="toc-Export-given-positions-into-a-table-format_002e" href="OutputAsTable.html#OutputAsTable">2.4.3 Export given positions into a table format.</a></li>
      <li><a name="toc-Call-SNPs-from-alignment-blocks-and-export-to-VCF-file" href="VcfOutput.html#VcfOutput">2.4.4 Call SNPs from alignment blocks and export to VCF file</a></li>
      <li><a name="toc-Export-segregating-sites-to-MSMC-format" href="MsmcOutput.html#MsmcOutput">2.4.5 Export segregating sites to MSMC format</a></li>
      <li><a name="toc-Export-segregating-sites-to-PLINK-format" href="PlinkOutput.html#PlinkOutput">2.4.6 Export segregating sites to PLINK format</a></li>
      <li><a name="toc-Output-coordinates-of-a-set-of-species-for-each-block_002e" href="OutputCoordinates.html#OutputCoordinates">2.4.7 Output coordinates of a set of species for each block.</a></li>
      <li><a name="toc-Write-trees-to-a-file_002e" href="OutputTrees.html#OutputTrees">2.4.8 Write trees to a file.</a></li>
      <li><a name="toc-Write-distance-matrices-to-a-file_002e" href="OutputDistanceMatrices.html#OutputDistanceMatrices">2.4.9 Write distance matrices to a file.</a></li>
    </ul></li>
    <li><a name="toc-Miscelaneous-filters" href="Miscelaneous.html#Miscelaneous">2.5 Miscelaneous filters</a>
    <ul class="no-bullet">
      <li><a name="toc-Convert-coordinates-from-one-genome-to-another" href="LiftOver.html#LiftOver">2.5.1 Convert coordinates from one genome to another</a></li>
      <li><a name="toc-Call-an-external-program-on-each-block_002e" href="SystemCall.html#SystemCall">2.5.2 Call an external program on each block.</a></li>
    </ul></li>
  </ul></li>
</ul>
</div>


<a name="Top"></a>
<div class="header">
<p>
Next: <a href="Introduction.html#Introduction" accesskey="n" rel="next">Introduction</a>, Previous: <a href="../dir/index.html" accesskey="p" rel="prev">(dir)</a>, Up: <a href="../dir/index.html" accesskey="u" rel="up">(dir)</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<hr>
<a name="The-MafFilter-Manual"></a>
<h1 class="top">The MafFilter Manual</h1>

<p>This is the manual of MafFilter, version 1.3.1.
</p>
<p>Copyright &copy; 2018 Julien Y. Dutheil
</p>







</body>
</html>
