<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<html>
<!-- This is the manual of MafFilter, version 1.3.1.

Copyright (C) 2018 Julien Y. Dutheil -->
<!-- Created by GNU Texinfo 6.1, http://www.gnu.org/software/texinfo/ -->
<head>
<title>MafFilter Manual 1.3.1: OutputAlignments</title>

<meta name="description" content="MafFilter Manual 1.3.1: OutputAlignments">
<meta name="keywords" content="MafFilter Manual 1.3.1: OutputAlignments">
<meta name="resource-type" content="document">
<meta name="distribution" content="global">
<meta name="Generator" content="makeinfo">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
<link href="index.html#Top" rel="start" title="Top">
<link href="index.html#SEC_Contents" rel="contents" title="Table of Contents">
<link href="Exporting.html#Exporting" rel="up" title="Exporting">
<link href="OutputAsTable.html#OutputAsTable" rel="next" title="OutputAsTable">
<link href="Output.html#Output" rel="prev" title="Output">
<style type="text/css">
<!--
a.summary-letter {text-decoration: none}
blockquote.indentedblock {margin-right: 0em}
blockquote.smallindentedblock {margin-right: 0em; font-size: smaller}
blockquote.smallquotation {font-size: smaller}
div.display {margin-left: 3.2em}
div.example {margin-left: 3.2em}
div.lisp {margin-left: 3.2em}
div.smalldisplay {margin-left: 3.2em}
div.smallexample {margin-left: 3.2em}
div.smalllisp {margin-left: 3.2em}
kbd {font-style: oblique}
pre.display {font-family: inherit}
pre.format {font-family: inherit}
pre.menu-comment {font-family: serif}
pre.menu-preformatted {font-family: serif}
pre.smalldisplay {font-family: inherit; font-size: smaller}
pre.smallexample {font-size: smaller}
pre.smallformat {font-family: inherit; font-size: smaller}
pre.smalllisp {font-size: smaller}
span.nolinebreak {white-space: nowrap}
span.roman {font-family: initial; font-weight: normal}
span.sansserif {font-family: sans-serif; font-weight: normal}
ul.no-bullet {list-style: none}
-->
</style>
<link rel="stylesheet" type="text/css" href="http://www.w3.org/StyleSheets/Core/Steely">


</head>

<body lang="en">
<a name="OutputAlignments"></a>
<div class="header">
<p>
Next: <a href="OutputAsTable.html#OutputAsTable" accesskey="n" rel="next">OutputAsTable</a>, Previous: <a href="Output.html#Output" accesskey="p" rel="prev">Output</a>, Up: <a href="Exporting.html#Exporting" accesskey="u" rel="up">Exporting</a> &nbsp; [<a href="index.html#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<a name="Write-alignment-blocks-to-an-external-alignment-file_002e"></a>
<h4 class="subsection">2.4.2 Write alignment blocks to an external alignment file.</h4>

<p>The <code>OutputAlignments</code> filter writes all blocks to an external alignment file, potentially losing some information such as coordinates and scores.
Sequence names will be formated following the Ensembl convention when coordinates are available.
</p>
<a name="Synopsis_003a-35"></a>
<h3 class="heading">Synopsis:</h3>

<table class="cartouche" border="1"><tr><td>
<div class="example">
<pre class="example">maf.filter=                                 \
    [...],
    OutputAlignments(                       \
        file=data.filtered.aln,             \
        compression=none,                   \
        format=Clustal,                     \
        mask=no,                            \
        coordinates=yes),                   \
    [...]
</pre></div>
</td></tr></table>

<a name="Arguments_003a-35"></a>
<h3 class="heading">Arguments:</h3>

<dl compact="compact">
<dt><code>file={none|{path}}</code></dt>
<dd><p>A file path where to write alignment blocks. If the file name contains the %i code, one alignment file per block will be generated, where %i will be replaced by the block index.
In addition to the %i special code, additional variables can be used (in combination with %i, in order to avoid putative identical file names): chromosome number (%c), begin (%b) and end (%e) of reference sequence.
</p>
</dd>
<dt><code>format={string}</code></dt>
<dd><p>One of the alignment format supported by Bio++ (this includes Fasta, Mase and Clustal, see the Bio++ Program suite reference manual for a complete list of formats and options).
</p>
</dd>
<dt><code>compression={none|gzip|zip|bzip2}</code></dt>
<dd><p>Compression format for output file.
</p>
</dd>
<dt><code>mask={yes|no}</code></dt>
<dd><p>Tell if sequences should be masked (if a mask annotation is available), or if masking information should be ignored.
</p>
</dd>
<dt><code>coordinates={yes|no}</code></dt>
<dd><p>Tell if coordinates of sequences should be written in the header of each sequence (followinf Ensembl syntax).
If set to no, or if sequences have no coordinates, only the species name will be used as a sequence name.
</p>
</dd>
<dt><code>ldhat_header={yes|no}</code></dt>
<dd><p>Tell if a header line should be added for each file, in order to be used with the LDhat convert program. This line contains the nimber of sequences, the sequence lengths, and the number 1 (for haploid).
</p>
</dd>
<dt><code>reference={string}</code></dt>
<dd><p>The name of reference sequence, which will only be used when %c, %b or %e codes are used in file names. In case a block does not contain a sequence for the reference species, &quot;ChrNA&quot;, &quot;StartNA&quot; and &quot;StopNA&quot; tags will be used.
</p>
</dd>
</dl>





</body>
</html>
